### PR TITLE
controller config caas-image-repo supports private or non dockerhub repos

### PR DESCRIPTION
**Pull request by ycliuhw**
This PR introduces non-dockerhub/private OCI registry support for:

- controller pods;
- model operators;
- application operators;
- charm container's base image in sidecar  pods;

This PR currently enabled `basic` auth and `auth token` support.
The [v2 bear token](https://docs.docker.com/registry/spec/auth/token/) is still a `TODO` and will be added in the following PRs.

Note: this feature is currently hidden under featureflag: `private-registry`, but it's not testable without some hack because [this bug(it will be fixed soon)](https://bugs.launchpad.net/juju/+bug/1941055).

## Checklist

 - [ ] ~Requires a [pylibjuju](https://github.com/juju/python-libjuju) change~
 - [ ] ~Added [integration tests](https://github.com/juju/juju/tree/develop/tests) for the PR~
 - [ ] ~Added or updated [doc.go](https://discourse.jujucharms.com/t/readme-in-packages/451) related to packages changed~
 - [x] Comments answer the question of why design decisions were made

### private repo on non dockerhub
```console
$ cat tttt.json
{
    "serveraddress": "quay.io",
    "auth": "xxxxxx==",
    // "username": "a",
    // "password": "pwd",
    "repository": "quay.io/ycliuhw"
}

$ JUJU_DEV_FEATURE_FLAGS=private-registry juju bootstrap microk8s k1 --config "features=[private-registry]" --config caas-image-repo="'$(cat tttt.json)'"

$ mkubectl get all,pv,pvc,ing -o wide -ncontroller-k1
NAME                                 READY   STATUS    RESTARTS   AGE     IP            NODE                  NOMINATED NODE   READINESS GATES
pod/modeloperator-687cb75dd8-c5p8c   1/1     Running   0          2m51s   10.1.97.208   kelvinliu-m15-ryzen   <none>           <none>
pod/controller-0                     2/2     Running   2          3m48s   10.1.97.207   kelvinliu-m15-ryzen   <none>           <none>

NAME                         TYPE        CLUSTER-IP       EXTERNAL-IP   PORT(S)     AGE     SELECTOR
service/controller-service   ClusterIP   10.152.183.22    <none>        17070/TCP   3m51s   app.kubernetes.io/name=controller
service/modeloperator        ClusterIP   10.152.183.161   <none>        17071/TCP   2m51s   operator.juju.is/name=modeloperator,operator.juju.is/target=model

NAME                            READY   UP-TO-DATE   AVAILABLE   AGE     CONTAINERS      IMAGES                                    SELECTOR
deployment.apps/modeloperator   1/1     1            1           2m51s   juju-operator   quay.io/ycliuhw/jujud-operator:2.9.10.1   operator.juju.is/name=modeloperator,operator.juju.is/target=model

NAME                                       DESIRED   CURRENT   READY   AGE     CONTAINERS      IMAGES                                    SELECTOR
replicaset.apps/modeloperator-687cb75dd8   1         1         1       2m51s   juju-operator   quay.io/ycliuhw/jujud-operator:2.9.10.1   operator.juju.is/name=modeloperator,operator.juju.is/target=model,pod-template-hash=687cb75dd8

NAME                          READY   AGE     CONTAINERS           IMAGES
statefulset.apps/controller   1/1     3m48s   mongodb,api-server   quay.io/ycliuhw/juju-db:4.0,quay.io/ycliuhw/jujud-operator:2.9.10.1

NAME                                                        CAPACITY   ACCESS MODES   RECLAIM POLICY   STATUS   CLAIM                                STORAGECLASS        REASON   AGE     VOLUMEMODE
persistentvolume/pvc-30c9754b-9b51-4182-bb14-0ebca15a49f5   20Gi       RWO            Delete           Bound    controller-k1/storage-controller-0   microk8s-hostpath            3m48s   Filesystem

NAME                                         STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS        AGE     VOLUMEMODE
persistentvolumeclaim/storage-controller-0   Bound    pvc-30c9754b-9b51-4182-bb14-0ebca15a49f5   20Gi       RWO            microk8s-hostpath   3m48s   Filesystem

$ juju add-model t1

$ juju deploy snappass-test

$ juju deploy cs:~juju/mariadb-k8s-3

$ mkubectl -ncontroller-k1 get pod/controller-0 -o json | jq .spec.imagePullSecrets
[
  {
    "name": "juju-image-pull-secret"
  }
]

$ mkubectl -ncontroller-k1 get secret/juju-image-pull-secret -o json | jq -r '.data[".dockerconfigjson"]' | base64 --decode | jq
{
  "auths": {
    "quay.io": {
      "auth": "xxxxxx==",
      "serveraddress": "quay.io"
    }
  }
}

$ mkubectl -nt1 get pods snappass-test-0 mariadb-k8s-operator-0 modeloperator-56b469cd6-n8dxp -o json |
 jq '.items[].spec.imagePullSecrets'
[
  {
    "name": "juju-image-pull-secret"
  }
]
[
  {
    "name": "juju-image-pull-secret"
  }
]
[
  {
    "name": "juju-image-pull-secret"
  }
]

$ mkubectl -nt1 get secret/juju-image-pull-secret -o json | jq -r '.data[".dockerconfigjson"]' | base64 --decode | jq
{
  "auths": {
    "quay.io": {
      "auth": "xxxxxx==",
      "serveraddress": "quay.io"
    }
  }
}

$ mkubectl -nt1 get pod/snappass-test-0 -o json | jq '.spec.containers[].image'
"quay.io/ycliuhw/charm-base:ubuntu-20.04"
"registry.hub.docker.com/library/redis@sha256:27bea08330830c6e5efc456075d66eb1fb4b76e5518947e8c87556adf1df3e51"
"registry.hub.docker.com/benhoyt/snappass-test@sha256:32506b5ac7bde75b7a76bee003cf8c77f30fc936f446c8c52d457b8fe6fce96f"

$ mkubectl -nt1 get pod/mariadb-k8s-operator-0 -o json | jq '.spec.containers[].image'
"quay.io/ycliuhw/jujud-operator:2.9.10"

$ mkubectl -nt1 get pod/modeloperator-56b469cd6-n8dxp -o json | jq '.spec.containers[].image'
"quay.io/ycliuhw/jujud-operator:2.9.10"

$ mkubectl -ncontroller-k1 get pod/controller-0  -o json | jq '.spec.containers[].image'
"quay.io/ycliuhw/juju-db:4.0"
"quay.io/ycliuhw/jujud-operator:2.9.10"

$ JUJU_DEV_FEATURE_FLAGS=private-registry juju upgrade-controller --agent-stream=develop


```

### public repo on dockerhub
```console
$ juju bootstrap microk8s k1 --config caas-image-repo=ycliuhw && juju add-model t1

$ mkubectl -ncontroller-k1 get pod/controller-0 -o json | jq .spec.imagePullSecrets
null

$ mkubectl -ncontroller-k1 get secret/juju-image-pull-secret
Error from server (NotFound): secrets "juju-image-pull-secret" not found

$ mkubectl -nt1 get pods snappass-test-0 mariadb-k8s-operator-0 modeloperator-84c9b5c68-9259k -o json |
 jq '.items[].spec.imagePullSecrets'
null
null
null

$ mkubectl -nt1 get secret/juju-image-pull-secret
Error from server (NotFound): secrets "juju-image-pull-secret" not found

$ mkubectl -nt1 get pod/snappass-test-0 -o json | jq '.spec.containers[].image'

"ycliuhw/charm-base:ubuntu-20.04"
"registry.hub.docker.com/library/redis@sha256:27bea08330830c6e5efc456075d66eb1fb4b76e5518947e8c87556adf1df3e51"
"registry.hub.docker.com/benhoyt/snappass-test@sha256:32506b5ac7bde75b7a76bee003cf8c77f30fc936f446c8c52d457b8fe6fce96f"

$ mkubectl -nt1 get pod/mariadb-k8s-operator-0 -o json | jq '.spec.containers[].image'
"ycliuhw/jujud-operator:2.9.10"

$ mkubectl -nt1 get pod/modeloperator-56b469cd6-n8dxp -o json | jq '.spec.containers[].image'
Error from server (NotFound): pods "modeloperator-56b469cd6-n8dxp" not found

$ mkubectl -nt1 get pod/modeloperator-84c9b5c68-9259k -o json | jq '.spec.containers[].image'
"ycliuhw/jujud-operator:2.9.10"

$ mkubectl -ncontroller-k1 get pod/controller-0  -o json | jq '.spec.containers[].image'

"ycliuhw/juju-db:4.0"
"ycliuhw/jujud-operator:2.9.10"

```

## Documentation changes

Yes

## Bug reference

https://bugs.launchpad.net/juju/+bug/1935953
https://bugs.launchpad.net/juju/+bug/1935830

### CODE REVIEW

**Comment by jnsgruk** [comment]
This is looking promising, let's please make sure a section ends up in the docs for this! :)

**Comment by ycliuhw** [comment]
> This is looking promising, let's please make sure a section ends up in the docs for this! :)

Yes, That's in the plan.

**Comment by ycliuhw** [comment]
[Juju doesn't support authentication for caas-image-repo](https://trello.com/c/8iyTHlhq/1678-juju-doesnt-support-authentication-for-caas-image-repo)

**Comment by ycliuhw** [comment]
[juju upgrade-controller with custom caas-image-repo failed](https://trello.com/c/19kfiDDs/1676-juju-upgrade-controller-with-custom-caas-image-repo-failed)

**Comment by ycliuhw** [comment]
[juju upgrade-controller with custom caas-image-repo failed](https://trello.com/c/19kfiDDs/1676-juju-upgrade-controller-with-custom-caas-image-repo-failed)

File: docker/auth.go [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+
+func (ba BasicAuthConfig) Empty() bool {
+	return ba.Auth == "" && ba.Username == "" && ba.Password == ""
+}
+
+func (ba BasicAuthConfig) Validate() error {
+	return nil
+}
+
+type ImageRepoDetails struct {
```
Comment (wallyworld) [inline docker/auth.go:65]: should comment

**Review by wallyworld — APPROVED** [review]
Looks nice.
One issue that needs fixing before landing - the apiserver/params should not use structs from other packages. We need to define bespoke structs in params and use helper methods to marshall/unmarshall.

File: apiserver/facades/controller/caasmodelconfigmanager/facade.go [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+	ctrlSt ControllerState,
+) (*Facade, error) {
+	return &Facade{
+		backend:   backend,
+		resources: resources,
+		ctrlSt:    ctrlSt,
+	}, nil
+}
+
+// Watch returns a watcher that notifies controller config changes.
```
[the remark this answers is not in the document]
Comment (ycliuhw) [reply apiserver/facades/controller/caasmodelconfigmanager/facade.go]: this has been removed

File: apiserver/facades/controller/caasmodelconfigmanager/state.go [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+}
+
+// ControllerState provides the subset of controller state
+// required by the CAAS application facade.
+type ControllerState interface {
+	ControllerConfig() (jujucontroller.Config, error)
+}
+
+type stateShim struct {
+	st Backend
```
[the remark this answers is not in the document]
Comment (ycliuhw) [reply apiserver/facades/controller/caasmodelconfigmanager/state.go]: this has been removed

File: apiserver/facades/controller/caasoperatorprovisioner/provisioner.go [hunk 4]
```diff
@@ -177,10 +178,15 @@ func (a *API) OperatorProvisioningInfo(args params.Entities) (params.OperatorPro
 		modelConfig,
 	)
 
-	imagePath, err := podcfg.GetJujuOCIImagePath(cfg, vers.ToPatch(), version.OfficialBuild)
-	if err != nil {
+	imageDetails := resources.DockerImageDetails{}
+	imageDetails.ImageRepoDetails = cfg.CAASImageRepo()
```
[the remark this answers is not in the document]
Comment (ycliuhw) [reply apiserver/facades/controller/caasoperatorprovisioner/provisioner.go:182]: not quite understand

File: apiserver/facades/controller/caasapplicationprovisioner/provisioner.go [hunk 5]
```diff
@@ -260,6 +260,8 @@ func (a *API) provisioningInfo(appName names.ApplicationTag) (*params.CAASApplic
 	if err != nil {
 		return nil, errors.Trace(err)
 	}
+	imageRepo := cfg.CAASImageRepo()
+	imageRepo.ServerAddress = ""
```
[the remark this answers is not in the document]
Comment (ycliuhw) [reply apiserver/facades/controller/caasapplicationprovisioner/provisioner.go:264]: this facade does not use the serveraddress, so cleared the default `ServerAddress` for less confusion.
But this piece of code has been changed. so we just ignore here now.

File: caas/kubernetes/provider/modeloperator.go [hunk 6]
```diff
@@ -379,18 +381,18 @@ func modelOperatorDeployment(
...
-	operatorImagePath string,
+	operatorImageDetails coreresources.DockerImageDetails,
 	port int32,
 	modelUUID,
 	serviceName,
 	serviceAccountName string,
 	volumes []core.Volume,
 	volumeMounts []core.VolumeMount,
-) (*apps.Deployment, error) {
+) (o *apps.Deployment, err error) {
```
[the remark this answers is not in the document]
Comment (ycliuhw) [reply caas/kubernetes/provider/modeloperator.go:391]: o -> output :)

File: docker/docker_test.go [hunk 7]
```diff
@@ -1,7 +1,7 @@
 // Copyright 2019 Canonical Ltd.
 // Licensed under the AGPLv3, see LICENCE file for details.
 
-package docker
+package docker_test
```
[the remark this answers is not in the document]
Comment (ycliuhw) [reply docker/docker_test.go:4]: NewImageInfo is actually exported for other packages but not only for tests inside this package.
As the package is getting bigger, it'd probably be better to move the tests into docker_test package?

**Comment by ycliuhw** [comment]
!!build!!

**Comment by ycliuhw** [comment]
$$merge$$

**Comment by ycliuhw** [comment]
[integrate more registry support, for example gcr, azure, etc.](https://trello.com/c/K1isSDRz/1749-integrate-more-registry-support-for-example-gcr-azure-etc)

**Pull request merged**